### ISSUE TITLE
Relax a test to address poorer than expected performance on Windows?

### ISSUE THREAD

**Issue by ablaom** [JuliaAI/CategoricalDistributions.jl#60]
A Windows fail was detected for [a PR](https://github.com/JuliaAI/CategoricalDistributions.jl/pull/59) making a change apparently unrelated to the failing test. Here is the log:

https://github.com/JuliaAI/CategoricalDistributions.jl/actions/runs/6160604583/job/16728423043?pr=59#step:6:232

Instead of a 10x speedup in a test for array-optimised arithmetic, we're only getting 2x speed up. I guess this is because of some change in the virtual machine used by GitHub for windows testing, or similar?

It may be the fail in intermittent. 

In any case, a little investigation is probably warranted before relaxing the test.

**Comment by skleinbo** [comment]
I can reproduce the test regression locally on Mac.

<details><summary>versioninfo</summary>

```
julia> versioninfo()
Julia Version 1.9.3
Commit bed2cd540a1 (2023-08-24 14:43 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 4 on 4 virtual cores
Environment:
  JULIA_CONDAPKG_BACKEND = Current
  JULIA_EDITOR = code
  JULIA_NUM_THREADS = 4
```

</details> 

Runs fine on `#dev` (https://github.com/JuliaAI/CategoricalDistributions.jl/commit/4be9db4197606d03480fe110f148aa1053b6eb1f), but fails with #59.

Curiously, I cannot reproduce on Windows 10.

**Comment by ablaom** [comment]
Thanks @skleinbo for the further investigation. This is pretty weird. I'm wondering if we can reproduce this error on any Julia version prior to 1.9 .

**Comment by skleinbo** [comment]
@ablaom Both versions, pre- and post #59, pass for me on Julia 1.8.5. 

The poor performance isn't real by the way, but for some reason recompilation seems to be triggered. 

This is going to be impossible to debug I fear. I can make the test pass by putting literally any kind of additional print statement in that test set. For example, annotating 

https://github.com/JuliaAI/CategoricalDistributions.jl/blob/66f23a3d50ffc180d8d185a34034d5025972caba/test/arithmetic.jl#L60-L61

with `@show`, or even putting a completely unrelated `@info "Test"` somewhere will make the test pass. It even passes when I switch those two lines - no I am not kidding.
I haven't the slightest clue how any of this connects, or where to even start debugging this.

Is it a good ideal to rely on a single measurement of wall time anyway? This seems prone to spurious failure.
Instead these tests could use `@btime` from `BenchmarkTools`. Granted, it'll make the test suite run a couple seconds longer and adds a further test dependency, but seems solid.

**Comment by ablaom** [comment]
@skleinbo Thanks for the further analysis. BTW The problem is intermittent; see #62 . 

> Instead these tests could use @belapsed from BenchmarkTools. Granted, it'll make the test suite run a couple seconds longer and adds a further test dependency, but seems solid.

This is an excellent suggestion.

**Comment by skleinbo** [comment]
Okay, great! Please see #63.

**Comment by ablaom** [comment]
closed by #63
